qiskit_experiments/framework/transpilation.py

```diff
@@ -0,0 +1,217 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Functions for preparing circuits for execution
+"""
+
+from __future__ import annotations
+
+import importlib.metadata
+import logging
+from collections.abc import Sequence
+
+from qiskit import QuantumCircuit, QuantumRegister, transpile
+from qiskit.exceptions import QiskitError
+from qiskit.providers import Backend
+from qiskit.providers.options import Options
+from qiskit.pulse.calibration_entries import CalibrationPublisher
+from qiskit.transpiler import Target
+
+
+LOGGER = logging.getLogger(__file__)
+
+DEFAULT_TRANSPILE_OPTIONS = Options(optimization_level=0, full_transpile=False)
+if importlib.metadata.version("qiskit").partition(".")[0] != "0":
+    DEFAULT_TRANSPILE_OPTIONS["num_processes"] = 1
+
+
+def map_qubits(
+    circuit: QuantumCircuit,
+    physical_qubits: Sequence[int],
+    n_qubits: int | None = None,
+) -> QuantumCircuit:
+    """Generate a new version of a circuit with new qubit indices
+
+    This function iterates through the instructions of ``circuit`` and copies
+    them into a new circuit with qubit indices replaced according to the
+    entries in ``physical_qubits``. So qubit 0's instructions are applied to
+    ``physical_qubits[0]`` and qubit 1's to ``physical_qubits[1]``, etc.
+
+    This function behaves similarly to passing ``initial_layout`` to
+    :func:`qiskit.transpile` but does not use a Qiskit
+    :class:`~qiskit.transpiler.PassManager` and does not fill the circuit with
+    ancillas.
+
+    Args:
+        circuit: The :class:`~qiskit.QuantumCircuit` to re-index.
+        physical_qubits: The list of new indices for ``circuit``'s qubit indices.
+        n_qubits: Optional qubit size to use for the output circuit. If
+            ``None``, then the maximum of ``physical_qubits`` will be used.
+
+    Returns:
+        The quantum circuit with new qubit indices
+    """
+    if len(physical_qubits) != circuit.num_qubits:
+        raise QiskitError(
+            f"Circuit to map has {circuit.num_qubits} qubits, but "
+            f"{len(physical_qubits)} physical qubits specified for mapping."
+        )
+
+    # if all(p == r for p, r in zip(physical_qubits, range(circuit.num_qubits))):
+    #     # No mapping necessary
+    #     return circuit
+
+    circ_size = n_qubits if n_qubits is not None else (max(physical_qubits) + 1)
+    p_qregs = QuantumRegister(circ_size)
+    p_circ = QuantumCircuit(
+        p_qregs,
+        *circuit.cregs,
+        name=circuit.name,
+        metadata=circuit.metadata,
+        global_phase=circuit.global_phase,
+    )
+    p_circ.compose(
+        circuit,
+        qubits=physical_qubits,
+        inplace=True,
+        copy=False,
+    )
+    return p_circ
```

Comment:
Perhaps you need to set `._layout` since this is how the Qiskit qasm3 module distinguishes between physical and virtual circuits (our primary payload is qpy which doesn't care layout but in case someone prefers qasm3).

Reply:
Interesting. I was not familiar with that property. It does seem like it should be set. I would want to check that it is right to set that private attribute for this case.